rllib/algorithms/dt/dt_torch_policy.py

```diff
@@ -0,0 +1,586 @@
+        pred_action = preds[SampleBatch.ACTIONS][:, -1]
+
+        return pred_action, self.dist_class, []
+
+    @override(TorchPolicyV2)
+    def extra_action_out(
+        self,
+        input_dict: Dict[str, TensorType],
+        state_batches: List[TensorType],
+        model: TorchModelV2,
+        action_dist: TorchDistributionWrapper,
+    ) -> Dict[str, TensorType]:
+        # NOTE: this function is only and always called after action_distribution_fn
+        # as long as they're being called by TorchPolicyV2._compute_action_helper
+        # which is called by PolicyV2.compute_actions_from_input_dict.
+        # Since action_distribution_fn modifies input_dict, input_dict now contains
+        # the updated RETURNS_TO_GO. We extract it and outputs it as extra_action_out.
+        # This is used by env_runner and is actually how it adds custom keys to
+        # SimpleListCollector and allows ViewRequirements to work.
+        # This is also used in user inference in get_next_input_dict, which takes
+        # this output as one of the input.
+        return {
+            # rtg still has the leftover extra 3rd dimension for inference
+            SampleBatch.RETURNS_TO_GO: input_dict[SampleBatch.RETURNS_TO_GO][:, -1, 0],
+        }
+
+    @override(TorchPolicyV2)
+    def loss(
+        self,
+        model: ModelV2,
+        dist_class: Type[TorchDistributionWrapper],
+        train_batch: SampleBatch,
+    ) -> Union[TensorType, List[TensorType]]:
+        """Loss function.
+
+        Args:
+            model: The ModelV2 to run foward pass on.
+            dist_class: The distribution of this policy.
+            train_batch: Training SampleBatch.
+                Keys and shapes: {
+                    OBS: [batch_size, max_seq_len, obs_dim],
+                    ACTIONS: [batch_size, max_seq_len, act_dim],
+                    RETURNS_TO_GO: [batch_size, max_seq_len + 1, 1],
+                    TIMESTEPS: [batch_size, max_seq_len],
+                    ATTENTION_MASKS: [batch_size, max_seq_len],
+                }
+        Returns:
+            Loss scalar tensor.
+        """
+        train_batch = self._lazy_tensor_dict(train_batch)
+
+        # policy forward and get prediction
+        model_out, _ = self.model(train_batch)  # noop, just returns obs.
+        preds = self.model.get_prediction(model_out, train_batch)
+
+        # get the regression targets
+        targets = self.model.get_targets(model_out, train_batch)
+
+        # get the attention masks for masked-loss
+        masks = train_batch[SampleBatch.ATTENTION_MASKS]
+
+        losses = []
+
+        # action losses
+        if isinstance(self.action_space, Discrete):
+            action_loss = self._masked_cross_entropy_loss(
+                preds[SampleBatch.ACTIONS], targets[SampleBatch.ACTIONS], masks
+            )
+        elif isinstance(self.action_space, Box):
+            action_loss = self._masked_mse_loss(
```

Comment:
You have a lot of code repetition here. Here is my suggested update:
```
# in loss() call this 
loss = self._masked_loss(preds, targets, masks)
# fn definition
def _masked_loss(preds, targets, masks):
     for key in targets:
           assert key not in preds, 'for target {key} there is no prediction named from the output of the model'
           if self._is_discrete(key):
                loss += self.config[f'loss_coef_{key}'] * self._masked_xent_loss(preds[key], targets[key], masks)
           else:
                loss += self.config[f'loss_coef_{key}'] * self._masked_mse_loss(preds[key], targets[key], masks)

     return loss

def _is_discrete(self, key):
     return key == SampleBatch.Action and self.action_space.is_discrete
```

Reply:
and user can use the coefs to basically tell which output heads they should regress over.

Reply:
done

Reply:
You should create a unit-tests that tests whether setting the loss_coefficients properly is actually effective and not buggy.

Reply:
Added the tests

Reply:
awesome!